[RFC] Eclipse plugin

Hjdskes opened the pull request:
For real this time: an Eclipse plugin.

@Balletie @gfokkema Please provide early feedback on what is there currently!

Hjdskes committed:
TravisCI: build the Eclipse modules as well.

As explained in the aggregator pom, the Eclipse plugin module cannot be build
in the same reactor build (see [1]). Therefore, TravisCI needs to be configured
to first build the reactor normally and afterwards the Eclipse plugin module.

[1]: https://wiki.eclipse.org/Tycho/How_Tos/Dependency_on_pom-first_artifacts

Hjdskes committed:
Eclipse: provide a basic plugin

This plugin provides an implementation of IEditor, IDisplay, IMessageHook and
IResultHook. Since it's just a basic plugin to get early feedback, more advanced
features such as multiline input, history, syntax highlighting, error marking,
completions, et cetera are not yet implemented.

Currently, Java's Observer & Observable are used. I will see if this can be
converted into RxJava.

Hjdskes committed:
Core: make Repl an interface

The previous refactoring still did not allow the flexibility required for
different kinds of clients. Therefore, Repl has now been refactored into an
interface with the only method all implementations have in common: eval. Reading
input is now completely up to the client and not bound to any contract in any
way.

This also means that there cannot be an exit command in the core module anymore,
as every REPL is now substantially different than any other.

Hjdskes committed:
Move IEditor, IDisplay and IInputHistory back to core

The Eclipse implementation turned out to be similar, so the interfaces are
useful to have in core.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/ReplModule.java:
```diff
@@ -25,9 +26,11 @@
+ * {@link IResultHook}.
  */
-public class ReplModule extends SpoofaxModule {
+public abstract class ReplModule extends SpoofaxModule {
```
Instead of an abstract class we can also provide dummies for the interfaces. I think this is a better solution.

Balletie replied:
Dummies are only useful if they serve any purpose at all, which I think in this case they wouldn't.

Hjdskes replied:
Dummies can print warnings, just like the Spoofax dummies do. Example: https://github.com/metaborg/spoofax/blob/692ef4677d4a1b70494b0fc12f0c2b013408d247/org.metaborg.core/src/main/java/org/metaborg/core/project/DummyProjectService.java

Hjdskes replied:
In #29, the only dummy required would be an `IRepl` dummy. Note that a dummy can just be a lambda that prints an error, or better yet throws an exception so that Guice complains.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/IRepl.java:
```diff
@@ -0,0 +1,44 @@
+ * input. This {@link ICommandInvoker} invokes {@link IReplCommand}s, which use
+ * {@link org.metaborg.spoofax.shell.hooks} to display output.
+ */
+public interface IRepl {
```
Skip mentioned here that the only behavior to implement for a client is the `getInvoker` getter, which kind of defeats the purpose of an interface. I agree, but according to Gerlof having an abstract class at the top of your hierarchy is also not "nice", so I'm not sure which solution you prefer here.

Balletie replied:
It's fine for now I think, we shouldn't worry too much about it.

Hjdskes commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/client/IEditor.java:
```diff
@@ -15,10 +14,38 @@
+     *
+     * @return The prompt used by this EclipseEditor.
+     */
+    StyledText getPrompt();
```
I added this getter and the one below such that the `EclipseRepl` class has access to the prompts when it needs to append the input to the `EclipseDisplay`.

Balletie commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/IRepl.java:
```diff
@@ -0,0 +1,44 @@
+ *
+ * The default {@link #eval(String)} implementation uses an {@link ICommandInvoker} to process user
+ * input. This {@link ICommandInvoker} invokes {@link IReplCommand}s, which use
+ * {@link org.metaborg.spoofax.shell.hooks} to display output.
```
nice comment

Hjdskes commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/impl/TerminalUserInterface.java:
```diff
@@ -81,27 +81,49 @@ protected void saveLine(String lastLine) {
+                // Clear the lines for next input.
+                lines.clear();
+            }
+        } catch (IOException e) {
```
I removed the `throws` from the interface because not all implementations deal with `IOException`s when reading; it is therefore a burden to have this for all implementations.

Unia commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/impl/ConsoleReplModule.java:
```diff
@@ -26,11 +27,17 @@
+    @Override
+    protected void configureCommands() {
+        super.configureCommands();
+        commandBinder.addBinding("exit").to(ExitCommand.class).in(Singleton.class);
```
The `ExitCommand` is now tied to the implementation of `IRepl`, therefore, the current `ExitCommand` has been moved to the console module.

Balletie commented on org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/core/ReplTest.java:
```diff
@@ -1,75 +0,0 @@
- * Tests the REPL by simulating user input.
- */
-@RunWith(MockitoJUnitRunner.class)
-public class ReplTest {
```
Always sad to see a test go :(

Unia replied:
The lines it was supposed to test are also gone ;)

Unia commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/impl/ConsoleRepl.java:
```diff
@@ -35,19 +37,42 @@
             String input;
             setRunning(true);
-            while (running && (input = read()) != null) {
+            while (running && (input = editor.getInput()) != null) {
```
Because of the removal of `IOException` in the `IEditor` interface, the private method below is not needed anymore.

Hjdskes commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/impl/hooks/EclipseMessageHook.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * An Eclipse-based implementation of {@link IMessageHook}.
+ */
+public class EclipseMessageHook implements IMessageHook {
```
The two hooks implemented here are identical to the implementations in the console module; perhaps these should be moved to the core module in one way or another.

Balletie replied:
Yes, if they can then they should I think.

Hjdskes replied:
See https://github.com/spoofax-shell/spoofax-shell/pull/24#discussion_r65898910.

Balletie commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/ReplView.java:
```diff
@@ -0,0 +1,59 @@
+
+    private void setupDisplay(Injector injector) {
+        EclipseDisplay display = injector.getInstance(EclipseDisplay.class);
+        display.getControl().setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
```
why is this not done in the `EclipseDisplay` constructor?

Hjdskes replied:
Because it is not up to the specific widgets to position themselves.

Balletie replied:
fine

Balletie commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/ReplView.java:
```diff
@@ -0,0 +1,59 @@
+
+    private void setupEditor(Injector injector) {
+        this.editor = injector.getInstance(EclipseEditor.class);
+        this.editor.getControl().setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, false));
```
why is this not done in the `EclipseEditor` constructor?

Hjdskes replied:
Because it is not up to the specific widgets to position themselves.

Balletie commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/impl/TerminalUserInterface.java:
```diff
@@ -81,27 +81,49 @@ protected void saveLine(String lastLine) {
+                lines.clear();
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```
How can we handle this exception better?

Hjdskes replied:
I assume this exception occurs when the streams have been closed, in which case we can't do anything but exit. We could print the message to the user though.

Balletie replied:
a stack trace is not very user-friendly so we should change it to something else

Balletie commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/impl/EclipseEditor.java:
```diff
@@ -0,0 +1,112 @@
+        this.input = new Text(parent, SWT.SINGLE);
+        this.input.addKeyListener(this);
+
+        this.setPrompt(new StyledText(Color.GREEN, "=> "));
```
prompt should maybe be the same width of the continuation prompt.

Unia commented:
@gfokkema Also please look at the folded comments.

gfokkema commented:
Might want to consider passing Composite by assisted inject, instead of introducing state in the guice module.

Hjdskes committed:
Eclipse: use RxJava instead of Java's Observ{er,able}

Unia commented:
The RxJava implementation (6dc1d75545e5ed82a13dbe8bb799c52db7e5753b) is not really where we want to be, but before I improve that I want to have drawn out the model of how the `KeyListener`s are going to work (I'm thinking strategy pattern here with a listener for single line and a listener for multiline mode, but perhaps this is overkill -- I haven't really thought about the differences in keybindings yet).

Before I do that, however, I'm going to look into syntax highlighting as I think that's more important for the demo tomorrow.

gfokkema committed:
StyledText: store regions and text apart.

gfokkema committed:
StyledText: Use a wildcard, since we're not using the contents.

gfokkema committed:
ConsoleRepl: style more than just foreground color

Don't show ascii reset when there's no style elements, ie. set the style
to null then.

Hjdskes committed:
Eclipse: rewrite the UI of the plugin

Hjdskes committed:
Eclipse: properly schedule hooks on the UI thread

As noted in the Javadoc for the hooks: since hooks are called automatically by
the commands, which in turn are called by the REPL running in its own thread,
hooks try to update the UI from another thread than the UI thread. The new UIJob
implementation actually under the hood calls asyncExe like we did before, but it
has a bit more monitoring going on as well.

Hjdskes committed:
Eclipse: provide note on threading in JavaDoc

Hjdskes committed:
Eclipse: don't use a default color

Hjdskes committed:
IDisplay: accept ISpoofaxResults

This allows client access to the full result, including error messages and
region and the like to display things however they need.

Hjdskes committed:
StyledText: merge text and regions again

If our commands set the text correctly in the ISourceRegions, there is no reason
anymore for these to be separate.

Unia commented:
@Balletie @gfokkema New features such as completions, hovers, syntax highlighting et cetera require more changes than I'm willing to make in this PR. I think it's ready to review: it provides a basic plugin that is already usable.

Balletie commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/impl/hooks/EclipseMessageHook.java:
```diff
@@ -26,13 +34,16 @@ public EclipseMessageHook(IDisplay display) {
+        Job job = new UIJob("REPL Message Hook") {
             @Override
-            public void run() {
+            public IStatus runInUIThread(IProgressMonitor monitor) {
```
This is more or less the same code three times in a row. Can this be fixed?

Hjdskes replied:
Sure, if we can figure out a nice way to do this. However, in #29 the hooks have been reworked so that a hook is returned to the client. This allows the client to schedule the hook however it needs, so we can schedule all hooks on the UI thread in one location -- so perhaps, we shouldn't look into resolving this here but do it in #29 instead.

Hjdskes committed:
Console: move tests to the correct package

Hjdskes committed:
Eclipse: clean up MANIFEST.MF

Hjdskes committed:
Eclipse: clean up pom.xml

gfokkema commented on org.metaborg.spoofax.shell.console/src/test/java/org/metaborg/spoofax/shell/client/console/impl/ConsoleReplTest.java:
```diff
@@ -85,10 +84,10 @@ private void setUp(String inputString) throws UnsupportedEncodingException {
     private void setUpCtrlD() throws IOException {
         setUp(String.valueOf(C_D));
-        invokerMock = mock(ICommandInvoker.class, RETURNS_MOCKS);
+        invokerSpy = mock(ICommandInvoker.class, RETURNS_MOCKS);
```
lolwut, spy = mock(..)

Hjdskes replied:
Oops, will change that back tomorrow.

Hjdskes replied:
Ah, I see now why I changed this. In this test, it is indeed a mock. In `setUpExit()`, it is a spy. Anyway, everything is now just named `mock` again.

gfokkema commented on org.metaborg.spoofax.shell.eclipse/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,23 @@
+Bundle-Vendor: metaborg.org
+Export-Package: org.metaborg.spoofax.shell.client.eclipse,
+ org.metaborg.spoofax.shell.client.eclipse.impl,
+ org.metaborg.spoofax.shell.client.eclipse.impl.history,
```
Package does not exist.

Hjdskes replied:
Will remove it tomorrow.

gfokkema commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/Activator.java:
```diff
@@ -0,0 +1,58 @@
+    @Override
+    public void start(BundleContext context) throws Exception {
+        super.start(context);
+        Activator.plugin = this;
```
Does this need a check to verify plugin = null, before assigning?

Hjdskes replied:
No, it's started only once by Eclipse.

Unia commented:
I have resolved all of @gfokkema's comments. @Balletie your turn :smile:

Hjdskes committed:
Core: move IRepl and ReplModule to the client package

Unia commented:
Rebased on #20.

Hjdskes committed:
TravisCI: work around a bug in the Eclipse compiler

As such, the aggregator pom is now redundant and removed.

Unia commented:
After 292c02c5402c0490dba3b8d36947c57a78fee6a3, the [aggregator pom](https://github.com/spoofax-shell/spoofax-shell/blob/eclipse/pom.xml) is completely redundant. Perhaps we should remove it?

Hjdskes committed:
Resolve external dependencies required for the shell

These were removed from `org.metaborg.spoofax.eclipse.externaldeps`[0, 1], but
we need them.

[0]: https://github.com/metaborg/spoofax-eclipse/commit/d6f6ffa82b1a9e8b9081ced75a07431b67f4a038
[1]: https://github.com/metaborg/spoofax-eclipse/commit/f4edd1a4e5fc42a17debc6ec4bd34f7480a5d5b6

Hjdskes committed:
Core: remove @gfokkema's personal maven repository

Our ShellFacet went upstream -- @gfokkema's repository is no longer required.

Unia commented:
The build errors seem to have been resolved. It builds locally in Maven, for me. To actually run the Eclipse plugin from Eclipse, you need to open `org.metaborg.spoofax.eclipse.externaldeps`, `org.metaborg.spoofax.eclipse.meta` and `org.metaborg.meta.lang.dynsem.interpreter`. Next, go to Run -> Run Configurations, click on the Plug-ins tab and make sure to de-select `org.metaborg.spoofax.eclipse.externaldeps` and `org.metaborg.spoofax.eclipse.meta` under Target Platform. Make sure all three projects mentioned are selected under Workspace.

The Eclipse projects are to resolve the changes I made in our [Spoofax Eclipse fork](https://github.com/spoofax-shell/spoofax-eclipse/commits/spoofax-shell-fixes) which in due time will have to make it upstream. The dynsem project should be embedded into our `org.metaborg.spoofax.shell.eclipse.externaldeps` module, but for some reason this still does not work. Maven has no problem resolving it, so opening this project is purely an Eclipse-workaround.
